Propositional Logic Questions

AbcAeffchen opened the pull request:
This PR contains generators for questions about satisfiability and normal forms. It also contains the necessary infrastructure to handle propositional logic expressions.

Other things in this PR:
- extend the random class to provide random booleans, some other handy functions, and some minor refactoring
- the sanity check for generators to handle invalid parameters was removed. This is already handled somewhere else and generators only get valid parameters.
- Multiple choice and text Questions can now show additional feedback.
- Text questions can now show typing aid buttons.

holgerdell reviewed (commented):
Das sieht großartig aus, vielen Dank.

Zum Parser, da werden wir ja sicher so einige brauchen. Ich hatte gehofft, dass wir uns die nicht immer selbst schreiben müssen, sondern uns automatisiert erzeugen lassen können - dadurch könnten wir auch leichter die Grammatik anpassen. Sowas wie das hier: https://chevrotain.io/docs/ oder https://peggyjs.org/, aber es gibt auch andere https://tomassetti.me/parsing-in-javascript/.

AbcAeffchen commented on front-end/src/components/ExerciseTextInput.tsx:
```diff
@@ -69,6 +72,19 @@ export function ExerciseTextInput({
+      userInputRef.current.value.slice(0, userInputRef.current.selectionStart ?? 0) +
+        text +
+        userInputRef.current.value.slice(userInputRef.current.selectionEnd ?? 0),
+    )
```
Yes this works but we lose the position of the cursor if the button is used to insert something that is not at the end. I tried `.setSelectionRange()` but this does not work for some reason. Maybe the default behavior should be to jump to the end, anyway?

holgerdell replied:
ah, that's annoying.

- Does `e.preventDefault()` help here?
- Is the cursor position really reset by the button click, or by changing the contents? If the latter, you might be able to store the current cursor position and restore it in the same function insertText.
- If all else fails, React probably needs to preserve and restore the cursor position using useState.

AbcAeffchen replied:
I tested a little bit more.
when I remove the `.focus()` call and just use `.setSelectionRange(x,x)` the caret is on the right position if the input field is focused manually e.g. using tab or by clicking the field.

What I tried but didn't work:
- add a `onFocus` event to the input field that calls `.setSelectionRange(x,x)` (caret appears at the end when focused using `.focus()`
- add a `caretPosition` to the state of the field and update that in addition to the `onFocus` event. 

So it feels like `.focus()` does something else than just trigger a focus event.

There is also inconsistent behavior between browsers without the focus call.
On firefox butons insert text at the end while chrome inserts text in the front.
This inconsistency is fixed by calling `.focus()` which puts the caret at the end.

To conclude this: I will add the `.focus()` call to turn back the focus to the input field and nothing else, at least for the moment.

AbcAeffchen commented on front-end/src/components/ExerciseTextInput.tsx:
```diff
@@ -106,15 +122,23 @@ export function ExerciseTextInput({
-      ) : (
-        <b className="text-2xl">{t("feedback.incorrect")}</b>
-      )
+      [
```
The alternative is to define `message` as an empty list, use the long form of `if` and push elements to the list accordingly. It is about the same size but probably more readable so I change it.

AbcAeffchen commented on front-end/src/components/ExerciseTextInput.tsx:
```diff
@@ -150,6 +175,11 @@ export function ExerciseTextInput({
         <Markdown md={question.bottomText} />
       </div>
+      {(question.typingAid ?? []).map((el, index) => (
+        <Button variant="secondary" key={`ta-${index}`} onClick={() => insertText(el.input)}>
```
So in my use-case the buttons are roughly quadratic but I'm not sure if this is always true. So I would not add a specific size.

Thanks for pointing out the missing aria-label. I will add it. For this I also add a locale entry `typing-aid.label: "insert \"{{0}}\""`.

AbcAeffchen commented on front-end/src/components/ExerciseTextInput.tsx:
```diff
@@ -150,6 +175,11 @@ export function ExerciseTextInput({
       <div className="py-5 text-slate-600 dark:text-slate-400">
         <Markdown md={question.bottomText} />
       </div>
+      {(question.typingAid ?? []).map((el, index) => (
```
Sure. Makes the `className="mx-1"` obsolete.

AbcAeffchen commented on shared/src/question-generators/propositional-logic/normalForms.ts:
```diff
@@ -0,0 +1,281 @@
+    freetext_feedback_unknown_variables:
+      "Ihre Antwort verwendet andere Variablen als der gegebene Ausdruck.",
+    freetext_feedback_no_normal_form: "Ihre Antwort ist keine {{0}}.",
+    freetext_feedback_parse_error: "Ihre Antwort ist kein gültiger aussagenlogischer Ausdruck.",
```
Ah, danke für den Hinweis. Wird geändert.

AbcAeffchen commented on shared/src/question-generators/propositional-logic/normalForms.ts:
```diff
@@ -0,0 +1,281 @@
+    // initialize the RNG so the question can be generated again
+    const random = new Random(seed)
+
+    const varNames = random.choice(variableNames).slice(0, <number>parameters.size)
```
typescript kann unsere `expectedParameters` und `parameters` nicht verbinden. das `<numbers>` ist nur ein typehint der sagt was da für ein type drin steht. Das wissen wir ja schon, weil es ein anderes system sicherstellt, dass die Generatoren nur bekommen was sie bestellt haben. Das `?? 2` soll nur einen default Wert festlegen für den Fall, dass gar kein wert übergeben wird. Ich weiß aber auch gar nicht ob  das überhaupt nötig ist. Evtl. kann man das auch ganz weg lassen, weil  die Generatoren immer ihr volles set an Parametern bekommen?

holgerdell replied:
Ich weiß auch nicht. Die parameters und expectedParameters gefallen mir noch nicht wirklich. Vielleicht wäre der saubere Weg, sowas wie zod zu benutzen, oder sowas ähnliches zu machen: https://zod.dev/?id=basic-usage

AbcAeffchen replied:
Ich sehe ehrlich gesagt das Problem nicht an dieser Stelle. Ich bin mir bewusst, dass ich typehint/casting zu viel verwende. Das liegt daran, ständig irgend etwas seinen type nicht kennt und mein reflex ist den zu annotieren, was aber mehr ist als ich eigentlich beabsichtige (ich versuche das zu bessern)

An dieser Stelle gibt es aber keine vernünftige Garantie, die typescript versteht, aber wenn wir wissen, dass wir ein System haben, das die erwarteten Eingaben garantiert (unit tests), dann schreiben wir einen typehint dran, damit typescript zufrieden ist und gut ist. Der typehint wird dann später sogar noch entfernt und steht nicht im JS code.

AbcAeffchen commented on shared/src/question-generators/propositional-logic/normalForms.ts:
```diff
@@ -0,0 +1,281 @@
+    } while (!satisfiabilityProperties.satisfiable || !satisfiabilityProperties.falsifiable)
+
+    const dnf = random.bool()
+    const normalForm = dnf ? expression.toDNF() : expression.toCNF()
```
die Normalform wird nur in der Multiple Choice Variante verwendet. Ich übergebe sie jetzt aber statt sie neu zu berechnen.

AbcAeffchen commented on shared/src/question-generators/propositional-logic/normalForms.ts:
```diff
@@ -0,0 +1,281 @@
+
+  const correctAnswerIndex = answers.indexOf(correctAnswerStr)
+
+  return <MultipleChoiceQuestion>{
```
Habe es gerade nochmal angeschaut. Es ist das gleiche wie `as` nur eine andere syntax.

In manchen generatoren steht so etwas wie `const question: type = {...}; return question;`. Ich wollte die variable los werden. Ich dachte ich hätte ausprobiert das Objekt direkt zurückzugeben und hätte dabei Probleme bekommen. Aber es geht auch so.
habe den Typehint also entfernt.

AbcAeffchen commented on shared/src/question-generators/propositional-logic/normalForms.ts:
```diff
@@ -0,0 +1,281 @@
+      if (isDNF ? answer.isDNF() : answer.isCNF)
+        return {
+          correct: false,
+          correctAnswer: "$" + getCorrectNormalFormStr(expression, isDNF) + "$",
```
Stimmt. das hatte ich refactored und die Stelle übersehen.

AbcAeffchen commented on shared/src/question-generators/propositional-logic/normalForms.ts:
```diff
@@ -0,0 +1,281 @@
+      return {
+        correct: true,
+      }
+    } catch (e) {
```
The parser can throw exceptions. The exceptions contain information what they expected and where which could be displayed in the feedback (but isn't at the moment). If the try-catch block only contains the parse-call, `answer` cannot be `const` and needs to be declared outside.

holgerdell replied:
I see. I guess I prefer the pattern where the parser is always safe, that is, instead of throwing an exception it returns an error. However, let's not get bogged down with this too much, do whatever you feel is best.

AbcAeffchen replied:
I see. The reason I use an exception is that the parser is recursive and so I don't need to check the return value every time. The exception just bubbles all the way to the top. However, I changed it the following way: There is a helper function that makes the parser call, catches the exception and then returns either the syntax tree or the parser error. This has also the benefit that the generator does not have to handle the exception correctly. At the moment it catches every exception that might occur, but actually only the ParserError should be handled and everything else should be rethrown. If anything can happen anyway, I have not checked.

AbcAeffchen commented on shared/src/question-generators/propositional-logic/satisfiability.ts:
```diff
@@ -0,0 +1,155 @@
+  if (satisfiable === false) return t(translations, lang, "feedback_contradiction")
+  if (falsifiable === false) return t(translations, lang, "feedback_tautology")
+
+  const falsifiedBy = Object.keys(satisfiable)
```
Good catch. No idea what happened here. Fixed it.

AbcAeffchen commented on shared/src/utils/propositionalLogic.ts:
```diff
@@ -0,0 +1,848 @@
+export type SyntaxTreeNodeType = Operator | Literal
+
+function isOperator(obj: any): obj is Operator {
+  return (obj as Operator).type !== undefined
```
Seems so. I changed it. Not sure why I ended up using this other notation. I had some thing as a type before and `instanceof` does not work for types. Maybe this was the reason.

AbcAeffchen commented on shared/src/utils/propositionalLogic.ts:
```diff
@@ -0,0 +1,848 @@
+export function numToVariableValues(num: number, variableNames: string[]): VariableValues {
+  const values: VariableValues = {}
+  let i = 0
+  while (num) {
```
Ah, mir war nicht bewusst, dass `toString()` ein Argument bekommen kann. `num.tostring(2)` genügt auch schon. Aber ich bin nicht sicher ob ich das verwenden will. Ich muss ja eh durch jedes Bit iterieren.

AbcAeffchen commented on shared/src/utils/random.ts:
```diff
@@ -44,16 +52,57 @@ export default class Random {
    */
   int(min: number, max: number): number {
+    if (min > max) throw new Error("Value Error: min > max")
+    if (min === max) return min
```
Right, this case just bypasses generating a random float in [0,1), converting it to a float in [min, min+1) and then rounding it down to min, when we already know the result.
I can remove it if it bothers you.

holgerdell replied:
Kann auch so bleiben, ist ziemlich egal.

AbcAeffchen commented on shared/src/question-generators/propositional-logic/normalForms.ts:
```diff
@@ -0,0 +1,281 @@
+      const { variableNames: exprVars, truthTable: exprTruthTable } = expression.getTruthTable()
+      const { variableNames: answerVars, truthTable: answerTruthTable } = answer.getTruthTable()
+
+      if (exprVars !== answerVars)
```
You are right. Even if it works in this case, we should change it to something reliable.
We currently don't have a place for such generics like checking if two arrays are equal or something like that? So for the moment I would put it into the propositional logic file. Or have you some other place in mind?

holgerdell replied:
/shared/src/utils/ seems more appropriate. A standard library called `lodash` can do this: https://docs-lodash.com/v4/is-equal/
While you probably don't want to use this library, we could use their naming conventions. I do not have an opinion here.

AbcAeffchen replied:
ok, I will add a file `generics.ts` in `/shared/src/utils/` and try to follow the naming scheme of lodash. For the moment it looks like we do not need a lot of the functions. if this changes we can still decide to use lodash as a drop in replacement.

AbcAeffchen commented on shared/src/utils/random.ts:
```diff
@@ -73,6 +122,24 @@ export default class Random {
+   * @param size Size of the first array. If not provided a random number is chosen.
+   */
+  splitArray<T>(array: ReadonlyArray<T>, size?: number): Array<T>[] {
+    if (size === undefined) size = this.int(1, array.length - 1)
```
Adding a function to draw binomial distributed random values would be the right way I think. However, such a function would had to handle a wide variety of input sizes and on a first glance that requires to handle three cases. So for the moment I don't want to do this and just require the size.

holgerdell replied:
Ein einfacher Algorithmus wäre:
- Wähle ganze Zufallszahl $R$ zwischen $0$ und $2^n$
- Berechne das minimale $k\in\mathbb N$ mit $R \leq \sum_{i<=k} \binom{n}{k}$.
- Gib $k$ aus.

Für große $n,k$ geht es auch in erwarteter konstanter Zeit via Theorem 1.19 in https://publikationen.sulb.uni-saarland.de/bitstream/20.500.11880/25473/1/thesis_final.pdf

AbcAeffchen commented on shared/src/utils/propositionalLogic.ts:
```diff
@@ -0,0 +1,848 @@
+      : binaryOperatorTypes
+  const thisOperator = random.weightedChoice(
+    allowedOperators,
+    allowedOperators.length < binaryOperatorTypes.length
```
Das ist was letztlich dabei raus kommt. Allerdings ist `allowedOperators` entweder die volle Liste `binaryOperatorTypes` oder eins weniger. So hat man weniger magic numbers.

holgerdell replied:
Ja natürlich hast du recht, ich weiß nicht mehr, warum ich das komisch fand.

AbcAeffchen commented on shared/src/utils/random.ts:
```diff
@@ -44,16 +52,57 @@ export default class Random {
   }
 
+  /**
+   * Splits an input number into an array of random numbers that sum up to the input number
```
I renamed the function to `partition` and the comment now states the distribution. The intend was to get a uniform random partition with the given restrictions. However, the implementation was flawed, so I re-implemented the function and dropped the `max` parameter.

AbcAeffchen commented on shared/src/utils/random.ts:
```diff
@@ -97,14 +164,31 @@ export default class Random {
+    let elements: readonly T[]
+
+    if (weights === undefined) {
+      elements = (data as ReadonlyArray<[element: T, weight: number]>).map((w) => w[0])
```
Since `ReadonlyArray<>` is a type in typescript, there seems to be no way to check which type we actually received. This seems only to be possible for the primitive types, where `typeof` works, or for classes where `instanceof` works. For objects one has to rely on some unique properties that could be used to differentiate between object types. Neither works in this case. However, typescript already guaranties that the function is called either as `Readonly<[el, w]>, undefined` or as `Readonly<T>, number[]`. So this is where the type checking already happens. So I used `weights` as an indicator to decide which type data has.

Maybe a custom type guard like the following in more in the typescript spirit, than using `as`

```
function isSingleValueCall(_: any, weights: any): _ is ReadonlyArray<[element: T, weight: number]> {
  return weights === undefined
}
```
using this function typescript can determine the right type. But it is still the same check as before.

I changed it to use the function. I think it is more readable and shows the intend of the code. I also added the checks for empty data and a check if each value has a weight and throw an error if necessary.

holgerdell commented:
@AbcAeffchen
I found two bugs, see screenshots. I added a test case for variable names such as x_1.

<img width="604" alt="Screenshot 2024-05-04 at 10 33 50" src="https://github.com/goethe-tcs/algo-learn/assets/467425/0cae0d55-7d9f-4eef-8750-20f154ec77a1">
<img width="604" alt="Screenshot 2024-05-04 at 10 32 42" src="https://github.com/goethe-tcs/algo-learn/assets/467425/9b0e6ff7-73d0-469d-bb3c-2b6d7b79ca1e">

AbcAeffchen commented:
I fixed both bugs. The parser even recognizes more inputs now like `A\andB` (no space after `\and`. There was also a bug that if your simplify your normal form so it uses not all variables, it was considered wrong. This is now also fixed and I added a test case.

Fabianstw commented on front-end/src/components/ExerciseMultipleChoice.tsx:
```diff
@@ -130,19 +130,22 @@ export function ExerciseMultipleChoice({
+          t("incorrect")
+        ),
       )
+      if (feedbackObject?.feedbackText) message.push(<Markdown md={feedbackObject.feedbackText} />)
```
Just a little problem, currently between `t("see.above")` and the `Markdown` component is no space or newline. The words are written together, which (for example) results in `See aboveThe expression can be falsified by`

AbcAeffchen replied:
Thanks. Should be fixed now.
